be/src/exec/pipeline/exchange/sink_buffer.h

```diff
@@ -44,46 +48,64 @@ class SinkBuffer {
-                    _in_flight_rpc_num--;
-                    _is_cancelled = true;
+                    _in_flight_rpc_num.fetch_sub(1, std::memory_order_release);
+                    {
+                        std::lock_guard<std::mutex> l(_mutex);
+                        _is_finishing = true;
+                    }
                     LOG(WARNING) << " transmit chunk rpc failed";
                 });
                 closure->addSuccessHandler([this](const PTransmitChunkResult& result) noexcept {
-                    _in_flight_rpc_num--;
+                    _in_flight_rpc_num.fetch_sub(1, std::memory_order_release);
                     Status status(result.status());
                     if (!status.ok()) {
-                        _is_cancelled = true;
+                        {
+                            std::lock_guard<std::mutex> l(_mutex);
+                            _is_finishing = true;
+                        }
                         LOG(WARNING) << " transmit chunk rpc failed, " << status.message();
                     }
                 });
                 _closures[dest_instance_id] = closure;
-
                 _buffers[dest_instance_id] = std::queue<TransmitChunkInfo>();
+                _request_seqs[dest_instance_id] = 0;
             }
-        }
 
-        try {
-            _thread = std::thread{&SinkBuffer::process, this};
-        } catch (const std::exception& exp) {
-            LOG(FATAL) << "[ExchangeSinkOperator] create thread: " << exp.what();
-        } catch (...) {
-            LOG(FATAL) << "[ExchangeSinkOperator] create thread: unknown";
+            _expected_eos = 0;
+            for (auto& [_, num] : _num_sinkers_per_dest_instance) {
+                _expected_eos += num;
+            }
         }
     }
 
     ~SinkBuffer() {
-        _is_finished = true;
-        _buffer_empty_cv.notify_one();
-        _thread.join();
-
-        // TODO(hcf) is_finish() unable to judge such situation that when process()
-        // pickup request from buffer and before transmitting through brpc.
-        // at this moment, _in_flight_rpc_num equals 0 and no closure is in flight
-        // but it is going to send packet. To handle this properly, we need to wait
-        // all the closure finish its io job
+        // Handle abnormal exit of SinkBuffer
+        bool abnormal_exit = false;
+        if (!is_finished()) {
+            std::lock_guard<std::mutex> l(_mutex);
+            abnormal_exit = true;
+            _is_finishing = true;
+        }
+        if (abnormal_exit) {
+            for (auto& [_, closure] : _closures) {
+                auto cntl = &closure->cntl;
+                brpc::Join(cntl->call_id());
+            }
+        }
+
+        // Wait for the io task to exit to avoid
+        // DeferOp accessing illegal memory
+        std::promise<void>* exit_promise = nullptr;
+        {
+            std::lock_guard<std::mutex> l(_mutex);
+            exit_promise = _io_task_exit_promise.get();
```

Comment:
promise::get, block introduced,

Reply:
_io_task_exit_promise is std::unique_ptr, get method just get raw pointer